### PR TITLE
Performant Bracket Pair Colorization

### PR DESCRIPTION
**Pull request by hediet**
Before with some extension (~10.000ms):

recording

After:

recording

image

(<1 ms, more than 10k times faster)

More ToDos: https://github.com/microsoft/vscode/issues/130767

`"editor.bracketPairColorization.enabled": true` will enable this new feature!

### CODE REVIEW

**Comment by Tyriar** [comment]
So this is getting built in instead of being in an extension? Have we thought about using dynamic imports for optional features like this so the bundle size doesn't keep growing?

**Comment by hediet** [comment]
> So this is getting built in instead of being in an extension?

Yes, see discussion [here](https://github.com/microsoft/vscode/issues/128465#issuecomment-879089188).

Properly exposing tokens to the extension host is a very hard problem - basically, the tokens would need to be recomputed by the extension host.
Synchronizing color information back to the editor is also very hard, as a single change can cause all following brackets to change their position & their color.

This implemenation in the editor core solves both problems by efficiently reusing the existing token information and synchronously providing the decorations when the view requests them.

> Have we thought about using dynamic imports for optional features like this so the bundle size doesn't keep growing?

That is a nice idea!
However, this implementation could potentially also drive the existing bracket pair matching, which could increase accuracy:

image

(In this case, I would rather match the two outer parentheses, as `[...]` should limit the scope of the inner `(`)

**Comment by KMFSousa** [comment]
This looks related to issue: https://github.com/microsoft/vscode/issues/96899

**Comment by hediet** [comment]
@rebornix I think this is ready for review. The last details took me quite a while.

I will iterate on the performance. I had it at 50ms for checker.ts, but now it is 160ms again after I made it more correct and supported nested languages.

I think the most important points for the review are:
* `editor.bracketPairColorization.enabled` defaults to false. It should be easy to verify that in this case, the new code does nothing - so when bracket pair colorization is turned off, there is absolutely no performance hit and everything should work as usual.
* This feature only changes very little existing code, so it does not add a lot of debt
* I suggest to review each commit individually.
* You don't need to check the feature itself for correctness, I will iterate on it.

Thanks!

File: src/vs/editor/common/model/DecorationProvider.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (bpasero) [inline src/vs/editor/common/model/DecorationProvider.ts:1]: I think we typically have lowercase file names.

File: src/vs/editor/common/model/bracketPairColorizer/immutableSet.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (bpasero) [inline src/vs/editor/common/model/bracketPairColorizer/immutableSet.ts:1]: Maybe move utilities that are editor agnostic into `base` with some tests`? I.e. we have `map.ts`.

File: src/vs/editor/common/model/DecorationProvider.ts (hunk 1 above)
Comment (hediet) [reply src/vs/editor/common/model/DecorationProvider.ts:1]: Right, I have to get used to that. Maybe I used the "Move To File" refactoring that names the file as the symbol that is getting moved.

**Comment by misolori** [comment]
In terms of color, here are the ones I would use:

For dark, I'd keep the existing colors you have, and for light I'd use the following colors:

- `#9a6d03`
- `#C433BF`
- `#0879BF`

CleanShot 2021-08-10 at 17 30 37@2x

And here are the color contrast ratios for each item on their respective backgrounds:

CleanShot 2021-08-10 at 17 28 13@2x

**Comment by hediet** [comment]
Thanks for the colors @misolori! However, now even I have troubles distinguishing the colors of the light theme.

image

**Comment by isidorn** [comment]
@misolori they look great, but as @hediet said: can we somehow increase there contrast relative to one another? Not to the background.

**Comment by misolori** [comment]
Well, if you go down the route of increasing contrast between the colors then you'll break the color contrast ratio between each color and the background. So not sure what we want to do here.

**Comment by hediet** [comment]
> Well, if you go down the route of increasing contrast between the colors then you'll break the color contrast ratio between each color and the background

But it works nicely for the dark theme!

**Comment by IllusionMH** [comment]
Agree that it would be great to have colors that easy to distinguish and spot if there are repeated characters.

It's easy to find a lot of `((` and even 25 `(((` in TS files of VS Code repo, but there are much more cases of `)))` and even ~217 of `))))`.

Also are there plans to allow users configure list of colors to use (if they prefer to have more that 3 colors) as simple setting (as in BPC2) or it would rely on fixed number of `editorBracketHighlight.foregroundN` theme colors?

P.S. Interesting edge case [`) }])\n }]`](https://github.com/microsoft/vscode/blob/2e23a7c3adb8628bdb71149b9031a9a22eb028f6/src/vs/workbench/contrib/workspaces/browser/workspaces.contribution.ts#L65-L66) 🤣

**Comment by hediet** [comment]
> Also are there plans to allow users configure list of colors to use

Unfortuantely, themes do not support an unlimited number of colors.

However, we plan to support up to 6 (*) bracket pair colors with 3 of them being set to `#00000000`, indicating that they are not set.
That way, these colors are themeable and you (or theme authors) can add more.
(*) or more if requested

File: src/vs/editor/common/config/editorOptions.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ import { USUAL_WORD_SEPARATORS } from 'vs/editor/common/model/wordHelper';
 import { AccessibilitySupport } from 'vs/platform/accessibility/common/accessibility';
 import { IConfigurationPropertySchema } from 'vs/platform/configuration/common/configurationRegistry';
 import { IJSONSchema } from 'vs/base/common/jsonSchema';
+import { isArray, isStringArray } from 'vs/base/common/types';
```
Comment (hediet) [inline src/vs/editor/common/config/editorOptions.ts:15]: Looks nice!

Comment (hediet) [inline src/vs/editor/common/config/editorOptions.ts:15]: Looks nice!

**Review by alexdima — APPROVED** [review]
:+1: Only P0 is to add a limit for super large files.

**Comment by hediet** [comment]
> 👍 Only P0 is to add a limit for super large files.

Limit is 50k lines with an average of 100 characters per line, so that it just accepts `checker.ts`.
Maybe the limit is to high and needs to be tweaked later.

**Comment by KROSF** [comment]
image

There are some problems with closing jsx tags

**Comment by IllusionMH** [comment]
@KROSF It's better to create separate issue as this is PR that already merged.
Looks like `/` is not colored and therefore whole ligature doesn't get color. Try to set `"editor.fontLigatures": false`

**Comment by sanket-bhalerao** [comment]
is `showHorizontalScopeLine` is going to be a part of this?

**Comment by IllusionMH** [comment]
@sanket-bhalerao better to discuss it in #131001

**Comment by carlosmgilp** [comment]
> ## How to change colors
> Up to 6 colors can be specified, according to the code in this PR.
> 
> ```json
> {
>   "workbench.colorCustomizations": {
>     "editorBracketHighlight.foreground1": "#896c74",
>     "editorBracketHighlight.foreground2": "#6c8289",
>     "editorBracketHighlight.foreground3": "#736c89",
>     "editorBracketHighlight.foreground4": "#74896c",
>     "editorBracketHighlight.foreground5": "#896c82",
>     "editorBracketHighlight.foreground6": "#89736c"
>   }
> }
> ```

Hello how are you.
I'm trying to customize the colors and it doesn't work for me.
Try this way and it doesn't work.
You could explain well how to do it.
Copy paste the example and it doesn't work either.
Thanks.

**Comment by gjsjohnmurray** [comment]
> I'm trying to customize the colors and it doesn't work for me.

@carlosmgilp did you try the instructions at https://code.visualstudio.com/docs/getstarted/themes#_customizing-a-color-theme ?

**Comment by carlosmgilp** [comment]
> > I'm trying to customize the colors and it doesn't work for me.
> 
> @carlosmgilp did you try the instructions at https://code.visualstudio.com/docs/getstarted/themes#_customizing-a-color-theme ?

Hello, how are you? Thank you for your answer.
Excuse my English, I only speak Spanish.
Well, I'm trying to do it like that, but if you look, when writing it doesn't show me the option of "editorBracketHighlight.foregroundn"
I don't know if I'm doing something wrong.

image

These are the errors it shows me.

image

Thanks.

**Comment by misolori** [comment]
@carlosmgilp you need to remove the extra brackets on line 96 & 105:

image

**Comment by carlosmgilp** [comment]
> @carlosmgilp you need to remove the extra brackets on line 96 & 105:

Hello brother, how are you? Thank you so much for your help and everyone's.
It works great now.
Thank you.

**Comment by carlosmgilp** [comment]
Is there a way to activate the colored line between each bracket.
Beforehand thank you very much.

image

**Comment by carlosmgilp** [comment]
Hello how are you.
I wanted to make an inquiry.
Because when I have part of my code commented, the parentheses are still shown in color.
Is that a bug or I have to configure something else.
I am using AutoLISP as code.

image

**Comment by hediet** [comment]
See https://github.com/microsoft/vscode/issues/132249.

**Comment by carlosmgilp** [comment]
Hi @hediet thanks for your help.
I will publish it in the other post.
Thank you.

**Pull request merged**